### PR TITLE
Tool/document

### PR DESCRIPTION
**Pull request by Gustavosande**
<sub>
By opening a pull request, I certify that I hold the intellectual property of the code I am submitting,
and I am granting the initial authors of WBO a perpetual, worldwide, non-exclusive, royalty-free, and irrevocable license to this code.
</sub>

### CODE REVIEW

File: client-data/tools/document/document.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                        ctx.drawImage(image, 0, 0, image.width * scale, image.height * scale);
+                        var dataURL = ctx.canvas.toDataURL("image/webp", 0.7);
+
+                        // Compressed file size as data url, approximately 1/3 larger than as bytestream
+                        size = dataURL.length;
+
+                        // attempt again with an image that is at least 10% smaller
+                        scale = scale * Math.sqrt(Math.min(
+                                0.9,
+                                Tools.server_config.MAX_DOCUMENT_SIZE / size
```
Comment (Gustavosande) [inline client-data/tools/document/document.js:53]: Codacy found an issue: ['Tools' is not defined.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

**Comment by Gustavosande** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 31
           

Complexity increasing per file
==============================
- client-data/tools/document/document.js  4
         

Complexity decreasing per file
==============================
+ server/sockets.js  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246&bid=19724356)

File: client-data/tools/document/document.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                    imgCount++;
+                };
+            };
+            // Tools.change(Tools.prevToolName);
+        });
+    }
+
+    function draw(msg) {
+        var aspect = msg.w/msg.h;
+        var img = Tools.createSVGElement("image");
```
Comment (Gustavosande) [inline client-data/tools/document/document.js:82]: Codacy found an issue: ['Tools' is not defined.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: server/sockets.js [hunk 3]
```diff
@@ -66,7 +67,7 @@ function socketConnection(socket) {
 
 	var lastEmitSecond = Date.now() / config.MAX_EMIT_COUNT_PERIOD | 0;
 	var emitCount = 0;
-	socket.on('broadcast', noFail(function onBroadcast(message) {
+	socket.on('broadcast', noFail(async function onBroadcast(message) {
```
Comment (Gustavosande) [inline server/sockets.js:70]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: client-data/tools/document/document.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+                                0.9,
+                                Tools.server_config.MAX_DOCUMENT_SIZE / size
+                        ));
+                    } while (size > Tools.server_config.MAX_DOCUMENT_SIZE);
+
+                    var msg = {
+                        id: uid,
+                        type: "doc",
+                        data: dataURL,
+                        size: size,
```
Comment (Gustavosande) [inline client-data/tools/document/document.js:61]: Codacy found an issue: [Expected property shorthand.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: client-data/tools/document/document.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+            reader.onload = function (e) {
+                // use canvas to compress image
+                var image = new Image();
+                image.src = e.target.result;
+                image.onload = function () {
+
+                    assert_count();
+
+                    var uid = Tools.generateUID("doc"); // doc for document
```
Comment (Gustavosande) [inline client-data/tools/document/document.js:33]: Codacy found an issue: ['Tools' is not defined.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: server/configuration.js [hunk 6]
```diff
@@ -26,6 +26,12 @@ module.exports = {
     /** Maximum value for any x or y on the board */
     MAX_BOARD_SIZE: parseInt(process.env['WBO_MAX_BOARD_SIZE']) || 65536,
 
+    /** Maximum size of uploaded documents default 1MB */
+    MAX_DOCUMENT_SIZE: parseInt(process.env['WBO_MAX_DOCUMENT_SIZE']) || 1048576,
+
+    /** Maximum number of documents allowed */
+    MAX_DOCUMENT_COUNT: parseInt(process.env['WBO_MAX_DOCUMENT_COUNT']) || 5,
```
Comment (Gustavosande) [inline server/configuration.js:33]: Codacy found an issue: [Always provide a base when using parseInt() functions](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: server/sockets.js [hunk 7]
```diff
@@ -96,6 +97,28 @@ function socketConnection(socket) {
 			return;
 		}
 
+		var boardData;
+		if (message.data.type === "doc") {
+			boardData = await getBoard(boardName);
+
+			if (boardData.existingDocuments >= config.MAX_DOCUMENT_COUNT) {
+				console.warn("Received too many documents");
```
Comment (Gustavosande) [inline server/sockets.js:105]: Codacy found an issue: [Unexpected console statement.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: client-data/tools/document/document.js [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            };
+            // Tools.change(Tools.prevToolName);
+        });
+    }
+
+    function draw(msg) {
+        var aspect = msg.w/msg.h;
+        var img = Tools.createSVGElement("image");
+        img.id=msg.id;
+        img.setAttribute("class", "layer-"+Tools.layer);
```
Comment (Gustavosande) [inline client-data/tools/document/document.js:84]: Codacy found an issue: ['Tools' is not defined.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: server/boardData.js [hunk 9]
```diff
@@ -228,7 +228,11 @@ BoardData.load = async function loadBoard(name) {
 	try {
 		data = await fs.promises.readFile(boardData.file);
 		boardData.board = JSON.parse(data);
-		for (id in boardData.board) boardData.validate(boardData.board[id]);
+		boardData.existingDocuments = 0;
+		for (id in boardData.board) {
+			boardData.validate(boardData.board[id]);
```
Comment (Gustavosande) [inline server/boardData.js:233]: Codacy found an issue: [Function Call Object Injection Sink](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: client-data/tools/document/document.js [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+                        // Compressed file size as data url, approximately 1/3 larger than as bytestream
+                        size = dataURL.length;
+
+                        // attempt again with an image that is at least 10% smaller
+                        scale = scale * Math.sqrt(Math.min(
+                                0.9,
+                                Tools.server_config.MAX_DOCUMENT_SIZE / size
+                        ));
+                    } while (size > Tools.server_config.MAX_DOCUMENT_SIZE);
```
Comment (Gustavosande) [inline client-data/tools/document/document.js:55]: Codacy found an issue: ['Tools' is not defined.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: client-data/tools/document/document.js [hunk 11]
```diff
@@ -0,0 +1,103 @@
+(function documents() { //Code isolation
```
Comment (Gustavosande) [inline client-data/tools/document/document.js:1]: Codacy found an issue: [Move the invocation into the parens that contain the function.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: client-data/tools/document/document.js [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    function draw(msg) {
+        var aspect = msg.w/msg.h;
+        var img = Tools.createSVGElement("image");
+        img.id=msg.id;
+        img.setAttribute("class", "layer-"+Tools.layer);
+        img.setAttributeNS(xlinkNS, "href", msg.data);
+        img.x.baseVal.value = msg['x'];
+        img.y.baseVal.value = msg['y'];
```
Comment (Gustavosande) [inline client-data/tools/document/document.js:87]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: client-data/tools/document/document.js [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+                        w: this.width * scale || 300,
+                        h: this.height * scale || 300,
+                        x: (100 + document.documentElement.scrollLeft) / Tools.scale + 10 * imgCount,
+                        y: (100 + document.documentElement.scrollTop) / Tools.scale + 10 * imgCount
+                        //fileType: fileInput.files[0].type
+                    };
+
+                    assert_count();
+
+                    draw(msg);
```
Comment (Gustavosande) [inline client-data/tools/document/document.js:71]: Codacy found an issue: ['draw' was used before it was defined.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: server/configuration.js [hunk 14]
```diff
@@ -26,6 +26,12 @@ module.exports = {
     /** Maximum value for any x or y on the board */
     MAX_BOARD_SIZE: parseInt(process.env['WBO_MAX_BOARD_SIZE']) || 65536,
 
+    /** Maximum size of uploaded documents default 1MB */
+    MAX_DOCUMENT_SIZE: parseInt(process.env['WBO_MAX_DOCUMENT_SIZE']) || 1048576,
```
Comment (Gustavosande) [inline server/configuration.js:30]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: client-data/tools/document/document.js [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+                        scale = scale * Math.sqrt(Math.min(
+                                0.9,
+                                Tools.server_config.MAX_DOCUMENT_SIZE / size
+                        ));
+                    } while (size > Tools.server_config.MAX_DOCUMENT_SIZE);
+
+                    var msg = {
+                        id: uid,
+                        type: "doc",
+                        data: dataURL,
```
Comment (Gustavosande) [inline client-data/tools/document/document.js:60]: Codacy found an issue: ['dataURL' used outside of binding context.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: client-data/tools/document/document.js [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+        img.setAttributeNS(xlinkNS, "href", msg.data);
+        img.x.baseVal.value = msg['x'];
+        img.y.baseVal.value = msg['y'];
+        img.setAttribute("width", 400*aspect);
+        img.setAttribute("height", 400);
+        Tools.drawingArea.appendChild(img);
+
+    }
+
+    Tools.add({
```
Comment (Gustavosande) [inline client-data/tools/document/document.js:94]: Codacy found an issue: ['Tools' is not defined.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: server/configuration.js (hunk 6 above)
Comment (Gustavosande) [inline server/configuration.js:33]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: client-data/tools/document/document.js [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+
+                    var msg = {
+                        id: uid,
+                        type: "doc",
+                        data: dataURL,
+                        size: size,
+                        w: this.width * scale || 300,
+                        h: this.height * scale || 300,
+                        x: (100 + document.documentElement.scrollLeft) / Tools.scale + 10 * imgCount,
+                        y: (100 + document.documentElement.scrollTop) / Tools.scale + 10 * imgCount
```
Comment (Gustavosande) [inline client-data/tools/document/document.js:65]: Codacy found an issue: ['Tools' is not defined.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: client-data/tools/document/document.js [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+        });
+    }
+
+    function draw(msg) {
+        var aspect = msg.w/msg.h;
+        var img = Tools.createSVGElement("image");
+        img.id=msg.id;
+        img.setAttribute("class", "layer-"+Tools.layer);
+        img.setAttributeNS(xlinkNS, "href", msg.data);
+        img.x.baseVal.value = msg['x'];
```
Comment (Gustavosande) [inline client-data/tools/document/document.js:86]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: client-data/tools/document/document.js [hunk 19]
```diff
@@ -0,0 +1,103 @@
+(function documents() { //Code isolation
+
+
+    var xlinkNS = "http://www.w3.org/1999/xlink";
+    var imgCount = 1;
+
+    function assert_count() {
+        if (Tools.svg.querySelectorAll("image").length >= Tools.server_config.MAX_DOCUMENT_COUNT) {
```
Comment (Gustavosande) [inline client-data/tools/document/document.js:8]: Codacy found an issue: ['Tools' is not defined.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: client-data/tools/document/document.js [hunk 20]
```diff
@@ -0,0 +1,103 @@
+(function documents() { //Code isolation
+
+
+    var xlinkNS = "http://www.w3.org/1999/xlink";
+    var imgCount = 1;
+
+    function assert_count() {
```
Comment (Gustavosande) [inline client-data/tools/document/document.js:7]: Codacy found an issue: [Identifier 'assert_count' is not in camel case.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: server/configuration.js (hunk 14 above)
Comment (Gustavosande) [inline server/configuration.js:30]: Codacy found an issue: [Always provide a base when using parseInt() functions](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: client-data/tools/document/document.js [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+                    } while (size > Tools.server_config.MAX_DOCUMENT_SIZE);
+
+                    var msg = {
+                        id: uid,
+                        type: "doc",
+                        data: dataURL,
+                        size: size,
+                        w: this.width * scale || 300,
+                        h: this.height * scale || 300,
+                        x: (100 + document.documentElement.scrollLeft) / Tools.scale + 10 * imgCount,
```
Comment (Gustavosande) [inline client-data/tools/document/document.js:64]: Codacy found an issue: ['Tools' is not defined.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: server/boardData.js (hunk 9 above)
Comment (Gustavosande) [inline server/boardData.js:233]: Codacy found an issue: ['id' is not defined.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

File: server/sockets.js [hunk 22]
```diff
@@ -96,6 +97,28 @@ function socketConnection(socket) {
...
+				return;
+			}
+
+			if (message.data.data.length > config.MAX_DOCUMENT_SIZE) {
+				console.warn("Received too large file");
```
Comment (Gustavosande) [inline server/sockets.js:110]: Codacy found an issue: [Unexpected console statement.](https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246)

**Comment by lovasoa** [comment]
These so-called "issues" completely miss the context of the code they highlight and some of them are just plain wrong.

**Comment by Gustavosande** [comment]
You can see more context here (https://app.codacy.com/manual/gustavosandesilva/whitebophir/pullRequest?prid=6017246&bid=19724356) on the issues found.

**Comment by lovasoa** [comment]
@Gustavosande : Almost none of these warnings is useful. You are doing no good to the product you are trying to promote here...

**Comment by Gustavosande** [comment]
Resolved.

**Pull request closed without merging**